src/libaktualizr/crypto/crypto.cc

```diff
@@ -341,14 +341,16 @@ StructGuard<EVP_PKEY> Crypto::generateRSAKeyPairEVP(KeyType key_type) {
       return {nullptr, EVP_PKEY_free};
   }
 
-#if AKTUALIZR_OPENSSL_PRE_11
-  StructGuard<RSA> rsa(RSA_generate_key(bits,     /* number of bits for the key - 2048 is a sensible value */
-                                        RSA_F4,   /* exponent - RSA_F4 is defined as 0x10001L */
-                                        nullptr,  /* callback - can be NULL if we aren't displaying progress */
-                                        nullptr), /* callback argument - not needed in this case */
-                       RSA_free);
-#else
   int ret;
+
+  ret = RAND_status();
```

Comment:
I've had a look at OpenSSL and I'm not convinced this is needed, as the same check is done in `rand_bytes`: https://github.com/openssl/openssl/blob/27007233db5d6f8b91ed474c4e09dd7014871cc6/crypto/rand/md_rand.c#L389.

A check like that only verifies that the library is self-consistent but the bad scenario is actually when openssl thinks everything was seeded properly but it read from `/dev/urandom`, blindly trusting that the kernel had finished seeding (urandom will never block).

After a quick strace, on recent distributions, openssl 1.1.1d uses `getrandom()` to seed its internal rng (safe). On xenial with Openssl 1.0.2g, it seems it's using urandom directly and uses the NONBLOCK + check for EAGAIN mentioned here https://lwn.net/Articles/606141/: https://github.com/openssl/openssl/blob/27007233db5d6f8b91ed474c4e09dd7014871cc6/crypto/rand/rand_unix.c#L377. I was afraid that it would read from urandom blindly which can be unsafe at early boot...

But maybe Openssl 1.0.0 with Linux >= 4.8 is unsafe, according to this recent comment? https://github.com/openssl/openssl/blob/0ab6fc79a9a63370be1a615729dc2a6ed0d6c89b/crypto/rand/rand_unix.c#L37

Though, we're most likely covered. If we're worried about the last point (to be confirmed), we could add a global blocking `getrandom()` followed by `RAND_poll()`

Reply:
Another interesting reference: https://github.com/openssl/openssl/pull/9595

Reply:
Hm actually I misread the LWN post and the EAGAIN check probably does not guarantee anything. The last solution might actually be the better one.

Reply:
> as the same check is done in rand_bytes

Erm, I'm kinda confused. The RAND_status() and RAND_poll() are not the same. Also I'd rather avoid depending on OpenSSL's internal implementation details. If RSA_generate_key_ex() man says that RNG should be seeded than I see no harm in explicitly checking that it's seeded - even in OpenSSL does the same internally.

> Maybe the simplest fix would be to have a globabl blocking getrandom()

This would introduce dependency on Linux kernel 3.17+  and glibc 2.25+ which is not available on xenial AFAIK. I'm also not sure how it differs from current PR - could you elaborate?

> bad scenario is actually when openssl thinks everything was seeded properly 

Indeed, my concern as well. On possible workaround could be using https://www.freedesktop.org/software/systemd/man/systemd-random-seed.service.html with entropy crediting to make sure we start after entropy pool has been taken care of.

Reply:
> > Maybe the simplest fix would be to have a globabl blocking getrandom()
> 
> This would introduce dependency on Linux kernel 3.17+ and glibc 2.25+ which is not available on xenial AFAIK. I'm also not sure how it differs from current PR - could you elaborate?

The syscall can also be used without going through glibc.
It differs from current PR because it would ensure that the systemd rng used by `/dev/urandom` is indeed seeded then force a reseed of openssl's RNG. Whereas your PR checks that openssl's RNG is seeded, though it may have used a bad source by reading `/dev/urandom` too early.